Add EIP: Identity aggregated NFT (#7231)
ethereum/EIPs: EIPS/eip-7231.md

---
eip: 7231
title: Identity aggregated NFT
description: The aggregation of web2 & web3 identities to NFTs, authorized by individuals, gives attributes of ownerships, relationships, experiences.

> author: Chloe Gu <chloe@carv.io>, Navid X. (@navid), Victor Yu <victor@carv.io>, Archer H.

Reviewer: You'll want to add your GitHub username (`xuxinlai2002`) here or else you won't be able to edit your own EIP.

Author: modify @navid -> @xuxinlai2002

> discussions-to: https://ethereum-magicians.org/t/eip-7231-namespaced-storage-layout/14796

Reviewer 2: Flagging this, the discussions-to link points to discussion on eip-7201 and not eip-7231.

@xuxinlai2002 might want to check this once.

Author: Thanks ,it has been revised.
https://ethereum-magicians.org/t/erc7231-identity-aggregated-nft/15062

status: Draft
type: Standards Track
category: ERC
created: 2023-06-25
requires: 165, 721, 1271, 6066
---

## Abstract

> This standard extends [ERC-721](./eip-721.md) by binding individuals' Web2 and Web3 identities to non-fungible tokens (SBTs included). By binding multiple identities, aggregated and composible identity infomation can be verified, resulting in more beneficial onchain scenarios for individuals, such as self-authentication, social overlapping, commercial value generation from user targetting, etc. 

Reviewer: It'd be nice to include a short _technical_ summary of the proposal in the abstract. You should answer the question of: how does this proposal attach identities to NFTs?

Author: add the technical summary  as follow：
By adding a custom schema in the metadata, and updating and verifying the schema hash in the contract, the binding of NFT and identity information is completed.

> This standard extends [ERC-721](./eip-721.md) by binding individuals' Web2 and Web3 identities to non-fungible tokens (SBTs included). By binding multiple identities, aggregated and composible identity infomation can be verified, resulting in more beneficial onchain scenarios for individuals, such as self-authentication, social overlapping, commercial value generation from user targetting, etc. By adding a custom schema in the metadata, and updating and verifying the schema hash in the contract, the binding of NFT and identity information is completed.

Reviewer: We generally prefer if abbreviations are expanded the first time they are used.

```suggestion
This standard extends [ERC-721](./eip-721.md) by binding individuals' Web2 and Web3 identities to non-fungible tokens (NFTs) and soulbound tokens (SBTs). By binding multiple identities, aggregated and composible identity infomation can be verified, resulting in more beneficial onchain scenarios for individuals, such as self-authentication, social overlapping, commercial value generation from user targetting, etc. By adding a custom schema in the metadata, and updating and verifying the schema hash in the contract, the binding of NFT and identity information is completed.
```

## Motivation

One of the most interesting aspects of Web3 is the ability to bring an individuals own identity to different applications. Even more powerful is the fact that individuals truly own their accounts without relying on centralized gatekeepers, disclosing to different apps components necessary for authentication and approved by individuals. 

> Exisiting solutions such as ENS, although open, decentralized, and make it more convenient for Ethereum-based applications, lack of data standardization and authentication of identity due to inherent anominity. Other solutions such as SBTs rely on centralized attestors, can not prevent data tampering, nor inscribe data into the ledger itself in a privacy enabling way. 

Reviewer 3: ```suggestion
Exisiting solutions such as ENS, although open, decentralized, and more convenient for Ethereum-based applications, suffer from a lack of data standardization and authentication of identity due to inherent anominity. Other solutions such as SBTs rely on centralized attestors, can not prevent data tampering, and do not inscribe data into the ledger itself in a privacy enabling way. 
```

Author: Thank you, it has been revised.

> Exisiting solutions such as ENS, although open, decentralized, and make it more convenient for Ethereum-based applications, lack of data standardization and athentication of itentity due to inherent anominity. Other solutions such as SBTs rely on centralized attestors, can not prevent data tampering, nor inscribe data into the ledger itself in a privacy enabling way. Previous work such as [ERC-6066](./eip-6066.md) does not bind nor verify user identity data, thus unable to meet the user identity aggregation, user Kyc verification and other application scenarios related to user data sovereignty

Author: Thanks，has been corrected

The proposed pushes the boundaries of solving identity problems with Identity Aggregated NFT, i.e., the individual-authenticated aggregation of web2 and web3 identities to NFTs (SBTs included). 

## Specification

The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY” and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

### Every compliant contract must implement the Interface

> ```typescript

Author: Thanks，has been corrected

    );

    /**
     * @notice 
     * @dev set the user ID binding information of NFT with identitiesRoot
     * @param id nft id 
     * @param identitiesRoot multi UserID Root data hash
     * MUST allow external calls
     */
    function setIdentitiesRoot(
        uint256 id,
        bytes32 identitiesRoot
    ) external;

    /**
     * @notice 
     * @dev get the multi-userID root by  NFTID
     * @param id nft id 
     * MUST return the bytes32 multiUserIDsRoot
     * MUST NOT modify the state
     * MUST allow external calls
     */
    function getIdentitiesRoot(
        uint256 id
    ) external returns(bytes32);

    /**
     * @notice 
     * @dev verify the userIDs binding 
    * @param id nft id 
     * @param userIDs userIDs for check
     * @param identitiesRoot msg hash to veriry
     * @param signature ECDSA signature 
     * MUST If the verification is passed, return true, otherwise return false
     * MUST NOT modify the state
     * MUST allow external calls
     */
    function verifyIdentitiesBinding(
        uint256 id,address nftOwnerAddress,string[] memory userIDs,bytes32 identitiesRoot, bytes calldata signature
    ) external returns (bool);
}
```

> This is the “ERCXXX Metadata JSON Schema” referenced above.

Reviewer: I don't see where this is mentioned above?

Author: Thanks，has been corrected
